Add WebSocket endpoint for chatbot and Todo.description

emmanuelbernard opened the pull request:
Will use WebSocket for bidirectional message to and from server
Added a description for Todo for the new UI

Here it is Phillip, I created an new field description for Todo
And created a WebSocket endpoint as /todo-agent/{todoId}

emmanuelbernard committed:
Add WebSocket endpoint for chatbot and Todo.description

Will use WebSocket for bidirectional message to and from server
Added a description for Todo for the new UI

emmanuelbernard commented:
@phillip-kruger ^ for the TODO UI

emmanuelbernard commented:
@cescoffier is that what you had in mind for the WebSocket?

cescoffier commented on src/main/java/io/quarkus/sample/TodoAgentWebSocket.java:
```diff
@@ -0,0 +1,34 @@
+            this.emitter = emitter;
+            // The emitter is now available for use elsewhere
+        });
+        emitter.emit("ok we will find agents for todo " + todoId );
```
Likely null, as the emitter will be created during the subscription.
So you can't emit at that point.

cescoffier replied:
Stream != Connection

You can immediately send a message back from the server to the client (just return what you want to send). 

But if ai understand correctly, you want the client to send a message with the todo id? In that case this todo is can be passed as path parameter or in a first message (handled using OnTextMessage).

cescoffier commented on src/main/java/io/quarkus/sample/TodoAgentWebSocket.java:
```diff
@@ -0,0 +1,34 @@
+    @OnTextMessage
+    void onTextMessage(@PathParam String todoId, String message) {
+        //do something with the agent context
+        emitter.emit("Parroting for " + todoId + " : " + message);
```
That will only work if you have a single connection otherwise the emitters will be mixed.

cescoffier reviewed (changes requested):
If you have an on message, just remove the onOpen and create the emitter in the onTextMessage.

cescoffier commented:
My advice when using web socket: See them as TCP connection. So on top of that you need to design and implement a protocol (that's why we often use json rpc on top of web socket in Quarkus as it provides a framing protocol). 

In your case I would create a few message type, with a 'kind' attribute to distinguish them.

emmanuelbernard committed:
Add agent protocol between fron and back

rsvoboda reviewed (commented):
Small typos in agent protocol file

phillip-kruger committed:
Added description and agent

Signed-off-by: Phillip Kruger <phillip.kruger@gmail.com>

emmanuelbernard committed:
Clarify possible cause of problem when AI fallback is used

emmanuelbernard committed:
Log AI calls and allow OpenAI key

emmanuelbernard committed:
Move emitter call in the emitter block to avoid NPE

emmanuelbernard committed:
Fix typos in protocol spec

phillip-kruger committed:
Add Json Message on backend

Signed-off-by: Phillip Kruger <phillip.kruger@gmail.com>

emmanuelbernard committed:
Create A2A client stub

emmanuelbernard committed:
Add test failing to subscribe on OnOpen

To test, go to UI, click on a TODO
then click on Create with AI green button
Then look at stack trace

emmanuelbernard committed:
Use Vert.X bus instead of Multi for communication across beans

This simplifies the subscription barrier issues

emmanuelbernard committed:
Add AgentSelector AI Service
Use CDI async event to pass info to AgentDispatcher
Create a Context holder for per todoId request after move to
@ApplicationScope

emmanuelbernard committed:
Finish support for end to end user to A2A agent server flow

emmanuelbernard committed:
Support A2A server not connected
Rename to AgentsMediator (from dispatcher)

emmanuelbernard committed:
Fix issue with user sent message in case we find no matching agent

emmanuelbernard committed:
Use AgentCard to populate the AgentSelector prompt

phillip-kruger committed:
Edit task edit feature

Signed-off-by: Phillip Kruger <phillip.kruger@gmail.com>

phillip-kruger commented:
https://github.com/user-attachments/assets/a4945fa0-c952-47fa-a9d1-03053e8b137d

emmanuelbernard committed:
Add second agent Movie Agent

Also adapt to different TAsk event behaviors
And ad more activity logs

emmanuelbernard committed:
Rename agent_request to agent_message and move packages for websocket classes

emmanuelbernard committed:
Support absent A2A servers and render AgentProducer more generic

emmanuelbernard committed:
Clean code for demo

emmanuelbernard committed:
Rename addTAskArtifact

emmanuelbernard committed:
Rename to A2AClient for clarity

emmanuelbernard committed:
Replace type if structure with switch case

emmanuelbernard committed:
Add mention of agentic in the readme

emmanuelbernard committed:
Clean code with switch cases
